File: JavaScript/200-number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        ROWS = len(grid)
        COLS = len(grid[0])

        visited = set()

        def dfs(r, c):
            dir = [(-1,0), (1,0), (0,-1), (0,1)]

            for i, j in dir:
                new_r = r + i
                new_c = c + j

                if new_r in range(ROWS) and new_c in range(COLS) and grid[new_r][new_c] == "1" and f'{new_r}-{new_c}' not in visited:
                    visited.add(f'{new_r}-{new_c}')
                    dfs(new_r, new_c)

        count = 0
        for i in range(ROWS):
            for j in range(COLS):
                if grid[i][j] == "1" and f'{i}-{j}' not in visited:
                    visited.add(f'{i}-{j}') 
                    dfs(i,j)
                    count+=1
        return count

File: JavaScript/200-number-of-islands/number-of-islands.java
class Solution {
    private int ROWS;
    private int COLS;
    private int islands = 0;
    private Set<String> visited = new HashSet<>();
    private char[][] grid;
    public int numIslands(char[][] grid) {
        ROWS = grid.length;
        COLS = grid[0].length;
        this.grid = grid;
        
        for(int i=0;i<ROWS;i++) {
            for(int j=0;j<COLS;j++) {
                String curr = String.valueOf(i)+ '_' + String.valueOf(j);
                if(this.grid[i][j] == '1' && !visited.contains(curr)) {
                    islands++;
                    dfs(i, j);
                }
            }
        }
        return islands;

    }

    public void dfs(int i, int j) {
        
        visited.add(String.valueOf(i)+ '_' + String.valueOf(j));

        int[][] directions = {{-1,0}, {1,0}, {0,-1}, {0,1}};

        for (int[] dir: directions){
            int new_r = i + dir[0];
            int new_c = j + dir[1];
            String curr = String.valueOf(new_r)+ '_' + String.valueOf(new_c);

            if(new_r >= 0 && new_r < ROWS && new_c >= 0 && new_c <COLS && !visited.contains(curr) && grid[new_r][new_c] == '1') {
                dfs(new_r, new_c);
            }
        }
    }
}